resources/assets/coffee/react/beatmaps.coffee

```diff
@@ -53,26 +58,64 @@ class @Beatmaps extends React.Component
   getSortState: =>
     'sort': [@state.sorting.field, @state.sorting.order].join('_')
 
-  search: =>
-    searchText = $('#searchbox').val().trim()
+  validSortFields: ->
+    ['title', 'artist', 'creator', 'difficulty', 'ranked', 'rating', 'plays']
+
+  validSortOrders: ->
+    ['asc', 'desc']
+
+  charToKey: (char) ->
+    switch char
+      when 'm' then 'mode'
+      when 's' then 'status'
+      when 'g' then 'genre'
+      when 'l' then 'language'
+      when 'e' then 'extra'
+      when 'r' then 'rank'
+      when 'q' then 'query'
 
+  search: =>
+    searchText = $('#searchbox').val()?.trim()
     @showLoader()
 
-    $.ajax(@state.paging.url,
-      method: 'get'
-      dataType: 'json'
-      data: $.extend({'q': searchText}, @getFilterState(), @getSortState())).done ((data) ->
-        more = data['data'].length > 10
-        @setState
-          beatmaps: data['data']
-          paging:
-            page: 1
-            url: @state.paging.url
-            loading: false
-            more: more
-          loading: false, ->
-            $(document).trigger 'beatmap:search:done'
-    ).bind(this)
+    # update url
+    filterState = @getFilterState()
+    sortState = @getSortState()
+
+    $.each filterState, (k,v) =>
+      if !v || @filterDefaults[@charToKey(k)] == v
+        delete filterState[k]
+
+    params = []
+    if searchText
+      params.push("q=#{searchText}")
+    for key, value of filterState
+      params.push("#{key}=#{value}")
+    for key, value of sortState
+      if value != 'ranked_desc'
+        params.push("so=#{value}")
+
+    if @state.just_restored or location.search.substr(1) != params.join('&')
+      if !@state.just_restored
+        history.pushState(@state, "¯\_(ツ)_/¯", "/beatmaps/?#{params.join('&')}")
+
+      $.ajax(@state.paging.url,
+        method: 'get'
+        dataType: 'json'
+        data: $.extend({'q': searchText}, @getFilterState(), @getSortState())).done ((data) ->
+          more = data['data'].length > 10
+          @setState
+            beatmaps: data['data']
+            query: searchText
+            paging:
+              page: 1
+              url: @state.paging.url
+              loading: false
+              more: more
+            just_restored: false
+            loading: false, ->
```

Comment:
Well, technically not...

But I'd just assign the whole thing to a variable first (or two for the callback) and pass that as the arguments.

Also that `more` variable doesn't seem to be useful.